torch/nn/modules/activation.py

```diff
@@ -691,8 +696,12 @@ def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=Fals
             self.q_proj_weight = Parameter(torch.Tensor(embed_dim, embed_dim))
             self.k_proj_weight = Parameter(torch.Tensor(embed_dim, self.kdim))
             self.v_proj_weight = Parameter(torch.Tensor(embed_dim, self.vdim))
+            self.register_parameter('in_proj_weight', None)
         else:
             self.in_proj_weight = Parameter(torch.empty(3 * embed_dim, embed_dim))
+            self.register_parameter('q_proj_weight', None)
```

Comment:
why we want to keep those None parameters around? Is this required by JIT?

Reply:
Yeah, the jit requires everything referenced on `self` in `forward` to be defined (even if it is define to be `None`). So even though the code will not execute it still needs to compile, hence the `None` parameters so the parameter definitions are the same for both paths (this is also the reason their names must be in `__constants__`)

Reply:
Let me ask the DDP people because they don't like unused parameters in general.

Reply:
For more data points this `register_parameter(..., None)` is used in many other modules including all the `Conv` modules, `BatchNorm`, `LayerNorm`, etc.